core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -741,7 +741,9 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
         longArray.set(pos * 2 + 1, keyHashcode);
         isDefined = true;
 
-        if (numKeys >= growthThreshold && longArray.size() < MAX_CAPACITY) {
+        // We use two array entries per key, so the array size is twice the capacity.
+        // We should compare the current capacity of the array, instead of its size.
+        if (numKeys >= growthThreshold && longArray.size() / 2 < MAX_CAPACITY) {
```

Comment:
@viirya . I agree your issue, but `longArray.size() / 2` might be too strict because we cap with `MAX_CAPACITY`. In other words, there is a room to grow still.
```
 // Allocate the new data structures
    allocate(Math.min(growthStrategy.nextCapacity(oldCapacity), MAX_CAPACITY));
```

Can we have explicit test cases for these boundary conditions?

Reply:
The max capacity is big number. Is it ok to have unit test allocating such big array?

Reply:
I guessed we can use `mock` with simple `growAndRehash` (although I didn't try).

Reply:
Never mind. I don't want to block this PR and you. I'll try that later by myself.

Reply:
Ok. Sounds good. I will also do test to see if I can add it. Thanks for the suggestion!

Reply:
hmm, even mock with `growAndRehash` to avoid allocate the array, `append` still needs to allocate memory page for values to insert. Mock with `append` makes less sense because it is where this logic remains.

Reply:
Oh, you already tested it. Got it. Thank you for spending time for that.